From c4c2525551eac6cf2ee79b89f42b744d6748405f Mon Sep 17 00:00:00 2001
Message-Id: <c4c2525551eac6cf2ee79b89f42b744d6748405f.1424394676.git.feitong.yi@intel.com>
In-Reply-To: <00f4430a265007e8e2c016cc7926bf530835ea30.1424394676.git.feitong.yi@intel.com>
References: <00f4430a265007e8e2c016cc7926bf530835ea30.1424394676.git.feitong.yi@intel.com>
From: Durgadoss R <durgadoss.r@intel.com>
Date: Mon, 2 Feb 2015 19:23:43 +0530
Subject: [PATCH 12/61] MUST_REBASE [VPG]: drivers/video/adf: Add PSR
 variables to dp_panel

This patch:
* Adds a psr_dpcd array to store PSR related DPCD register values.
* Print some useful information related to PSR during panel init.

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed

Issue: GMINL-5357
Change-Id: I1ee130deb3d18826066bf3c0d0f7722eafe06cd4
Signed-off-by: Durgadoss R <durgadoss.r@intel.com>
---
 drivers/video/adf/intel/core/common/dp/dp_panel.c  |   15 +++++++++++++++
 .../adf/intel/include/core/common/dp/dp_panel.h    |    3 +++
 2 files changed, 18 insertions(+)

diff --git a/drivers/video/adf/intel/core/common/dp/dp_panel.c b/drivers/video/adf/intel/core/common/dp/dp_panel.c
index 214130e..0a6aac43 100644
--- a/drivers/video/adf/intel/core/common/dp/dp_panel.c
+++ b/drivers/video/adf/intel/core/common/dp/dp_panel.c
@@ -394,6 +394,7 @@ err:
 bool dp_panel_init(struct dp_panel *panel, struct intel_pipeline *pipeline)
 {
 	int err = 0;
+	bool link_train_on_exit = true;
 
 	panel->pipeline = pipeline;
 	vlv_get_max_vswing_preemp(pipeline, &panel->max_vswing,
@@ -405,6 +406,20 @@ bool dp_panel_init(struct dp_panel *panel, struct intel_pipeline *pipeline)
 	pr_err("Received %d bytes for start panel %x %x\n", err,
 			panel->dpcd_start[0], panel->dpcd_start[1]);
 
+	/* Check if the panel supports PSR */
+	memset(panel->psr_dpcd, 0, sizeof(panel->psr_dpcd));
+	dp_panel_get_dpcd(panel, DP_PSR_SUPPORT,
+			panel->psr_dpcd, sizeof(panel->psr_dpcd));
+
+	if (panel->psr_dpcd[0] & DP_PSR_IS_SUPPORTED)
+		pr_info("Detected EDP PSR Panel.\n");
+
+	if (panel->psr_dpcd[1] & DP_PSR_NO_TRAIN_ON_EXIT)
+		link_train_on_exit = false;
+
+	pr_info("Link training is%srequired on PSR exit\n",
+			link_train_on_exit ? " " : " not ");
+
 	return true;
 }
 
diff --git a/drivers/video/adf/intel/include/core/common/dp/dp_panel.h b/drivers/video/adf/intel/include/core/common/dp/dp_panel.h
index d384d47..aa565f3 100644
--- a/drivers/video/adf/intel/include/core/common/dp/dp_panel.h
+++ b/drivers/video/adf/intel/include/core/common/dp/dp_panel.h
@@ -400,6 +400,9 @@ struct dp_panel {
 
 	/* First 12 bytes of dpcd read during init */
 	u8 dpcd_start[12];
+
+	/* DPCD bytes related to PSR [70-71h] */
+	u8 psr_dpcd[2];
 };
 
 extern bool dp_panel_init(struct dp_panel *panel,
-- 
1.7.9.5

